template/src/components/DeviceConfigure.tsx

```diff
@@ -12,35 +12,105 @@
 import React, {useState, useContext, useEffect, useCallback} from 'react';
 import {RtcContext, ClientRole} from '../../agora-rn-uikit';
 import DeviceContext from './DeviceContext';
+import AgoraRTC from 'agora-rtc-sdk-ng';
 
 interface Props {
   userRole: ClientRole;
 }
-
+interface deviceInfo {
+  deviceId: string;
+  groupId: string;
+  kind: string;
+  label: string;
+}
+interface changedDeviceInfo {
+  device: deviceInfo;
+  initAt: number;
+  state: 'INACTIVE' | 'ACTIVE';
+  updateAt: number;
+}
 const DeviceConfigure: React.FC<Props> = (props: any) => {
   const [selectedCam, setSelectedCam] = useState('');
   const [selectedMic, setSelectedMic] = useState('');
-  const [deviceList, setDeviceList] = useState([]);
+  const [deviceList, setDeviceList] = useState<any>([]);
   const rtc = useContext(RtcContext);
 
   const refreshDevices = useCallback(async () => {
-    rtc.RtcEngine.getDevices(function (devices: any) {
-      console.log('set devices');
-      setDeviceList(devices);
-      for (const i in devices) {
-        if (devices[i].kind === 'videoinput') {
-          setSelectedCam(devices[i].deviceId);
+    rtc.RtcEngine.getDevices(function (devices: deviceInfo[]) {
+      const uniqueDevices = devices
+        .filter(
```

Comment:
Can we comment this piece of code. It's hard to understand whats going on here

Reply:
@nitte93, comment added